accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -3556,6 +3499,20 @@ Here is one way it could be broken down into separate pieces:
 
 ## Changelog
 
+### 2.29
+
+-   Map patterns no longer check length.
+
+-   Remove `...` from map patterns since it is redundant with the previous
+    change.
+
+-   Make it an error to have an empty map pattern. Since map patterns don't
+    check their length, an empty map pattern will match all maps, which is
+    likely to confuse users. For now, to minimize confusion, we just disallow
+    it.
+
+-   Make it no longer an error for map patterns to have duplicate keys.
```

Comment:
Can't say what Bob was thinking, but it's a needless restriction if we don't actually try determine whether we hit all map keys.

When we tried to check that a map pattern with four sub-patterns matched all the entries of a map with four entries, it was important that no two keys were the same.
Now that we don't care about counting the patterns any more, it also doesn't matter if you test the same key twice. You can write `"a": var a, "a": List(isNotEmpty: true)` or `"a": var a && List(isNotEmpty: true)`, and it'll do the same thing.

So, basically, no need to make it an error, so why hit people over the head for doing something that's not actually wrong.

(Now, it might be *unnecessary* to have duplicate keys, but that's very different from being wrong. There might be lints against it, which heuristically detects whether it's reasonable or unreasonable, and whether two non-identical keys should be considered equivalent or not, something the language specification should not try to specify formally.)

Reply:
Lasse is correct. The original motivation for that restriction was:

> If a map pattern has multiple identical key entries, they will increase the required length for the pattern to match but in all but the most perverse `Map` implementations will represent the same key. Thus, it's very unlikely that any map pattern containing identical keys (and no rest element) will ever match. Duplicate keys are most likely a typo in the code.

It's still probably the case that duplicate keys is a typo, but having duplicate keys will no longer lead to unmatchable lists, so I figured it made sense to drop the restriction. For comparison, we allow duplicate field names in object patterns, and with this change, map patterns are most similar to object patterns in their behavior.

Reply:
OK I understand why we don't have to make it an error.
And it might be a lint.

BTW, I think we _don't_ allow duplicate fields in object patterns, the specification says "It is a compile-time error if: \<cut\> Any two named fields have the same name."

Reply:
Ah, crap. I completely misremembered that. :(

In that case, it probably makes sense to keep it an error for map patterns too. I'll start a thread.

Reply:
Also, thank you for pointing this out!